File: src/main/java/com/soapdogg/lispInterpreter/parser/AtomNodeParser.java
package com.soapdogg.lispInterpreter.parser;

import com.soapdogg.lispInterpreter.datamodels.Node;
import com.soapdogg.lispInterpreter.datamodels.ParserResult;
import com.soapdogg.lispInterpreter.datamodels.Token;
import com.soapdogg.lispInterpreter.generator.NodeGenerator;
import lombok.AllArgsConstructor;

import java.util.List;

@AllArgsConstructor(staticName = "newInstance")
public class AtomNodeParser {

    private final NodeGenerator nodeGenerator;
    private final ParserResultBuilder parserResultBuilder;

    public ParserResult parseAtomNode(
        final List<Token> tokens
    ) {
        final Token token = tokens.remove(0);
        final String value = token.getValue();
        final Node atomNode = nodeGenerator.generateAtomNode(value);
        return parserResultBuilder.buildParserResult(
            atomNode,
            tokens
        );
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/parser/AtomNodeParser.kt
package com.soapdogg.lispInterpreter.parser

import com.soapdogg.lispInterpreter.converter.NodeConverter
import com.soapdogg.lispInterpreter.datamodels.ParserResult
import com.soapdogg.lispInterpreter.datamodels.Token
import com.soapdogg.lispInterpreter.generator.NodeGenerator

class AtomNodeParser(
    private val nodeGenerator: NodeGenerator,
    private val nodeConverter: NodeConverter,
    private val parserResultBuilder: ParserResultBuilder
) {
    fun parseAtomNode(
        tokens: List<Token>
    ): ParserResult {
        val value = tokens[0].value
        val atomNode = nodeGenerator.generateAtomNode(value)
        val convertedAtomNode = nodeConverter.convertNodeV2ToNode(atomNode)
        return parserResultBuilder.buildParserResult(
            convertedAtomNode,
            tokens.subList(1, tokens.size)
        )
    }
}